File: src/main/java/ludwig/utils/ArgumentsCount.java
package ludwig.utils;

import ludwig.model.*;

import static ludwig.utils.NodeUtils.arguments;

class ArgumentsCount implements NodeVisitor<Integer> {
    @Override
    public Integer visitProject(ProjectNode projectNode) {
        return null;
    }

    @Override
    public Integer visitPackage(PackageNode packageNode) {
        return null;
    }

    @Override
    public Integer visitFunction(FunctionNode functionNode) {
        return Integer.MAX_VALUE;
    }

    @Override
    public Integer visitVariable(VariableNode variableNode) {
        return 0;
    }

    @Override
    public Integer visitAssignment(AssignmentNode assignmentNode) {
        return 2;
    }

    @Override
    public Integer visitReference(ReferenceNode referenceNode) {
        return arguments(referenceNode.ref()).size();
    }

    @Override
    public Integer visitList(ListNode listNode) {
        return Integer.MAX_VALUE;
    }

    @Override
    public Integer visitLiteral(LiteralNode literalNode) {
        return 0;
    }

    @Override
    public Integer visitLambda(LambdaNode lambdaNode) {
        return Integer.MAX_VALUE;
    }

    @Override
    public Integer visitCall(CallNode callNode) {
        return Integer.MAX_VALUE;
    }

    @Override
    public Integer visitReturn(ReturnNode returnNode) {
        return 1;
    }

    @Override
    public Integer visitIf(IfNode ifNode) {
        return Integer.MAX_VALUE;
    }

    @Override
    public Integer visitElse(ElseNode elseNode) {
        return Integer.MAX_VALUE;
    }

    @Override
    public Integer visitFor(ForNode forNode) {
        return Integer.MAX_VALUE;
    }

    @Override
    public Integer visitFunctionReference(FunctionReferenceNode functionReference) {
        return 1;
    }

    @Override
    public Integer visitThrow(ThrowNode throwNode) {
        return 1;
    }

    @Override
    public Integer visitPlaceholder(PlaceholderNode placeholderNode) {
        return 0;
    }

    @Override
    public Integer visitBreak(BreakNode breakNode) {
        return 1;
    }

    @Override
    public Integer visitContinue(ContinueNode continueNode) {
        return 1;
    }

    @Override
    public Integer visitOverride(OverrideNode overrideNode) {
        return arguments(overrideNode).size();
    }

    @Override
    public Integer visitClass(ClassNode classNode) {
        return arguments(classNode).size();
    }
}

File: src/main/kotlin/ludwig/utils/ArgumentsCount.kt
package ludwig.utils

import ludwig.model.*

import ludwig.utils.NodeUtils.arguments

internal class ArgumentsCount : NodeVisitor<Int?> {
    override fun visitProject(projectNode: ProjectNode): Int? {
        return null
    }

    override fun visitPackage(packageNode: PackageNode): Int? {
        return null
    }

    override fun visitFunction(functionNode: FunctionNode): Int? {
        return Integer.MAX_VALUE
    }

    override fun visitVariable(variableNode: VariableNode): Int? {
        return 0
    }

    override fun visitAssignment(assignmentNode: AssignmentNode): Int? {
        return 2
    }

    override fun visitReference(referenceNode: ReferenceNode): Int? {
        return arguments(referenceNode.ref).size
    }

    override fun visitList(listNode: ListNode): Int? {
        return Integer.MAX_VALUE
    }

    override fun visitLiteral(literalNode: LiteralNode): Int? {
        return 0
    }

    override fun visitLambda(lambdaNode: LambdaNode): Int? {
        return Integer.MAX_VALUE
    }

    override fun visitCall(callNode: CallNode): Int? {
        return Integer.MAX_VALUE
    }

    override fun visitReturn(returnNode: ReturnNode): Int? {
        return 1
    }

    override fun visitIf(ifNode: IfNode): Int? {
        return Integer.MAX_VALUE
    }

    override fun visitElse(elseNode: ElseNode): Int? {
        return Integer.MAX_VALUE
    }

    override fun visitFor(forNode: ForNode): Int? {
        return Integer.MAX_VALUE
    }

    override fun visitFunctionReference(functionReference: FunctionReferenceNode): Int? {
        return 1
    }

    override fun visitThrow(throwNode: ThrowNode): Int? {
        return 1
    }

    override fun visitTry(tryNode: TryNode): Int? {
        return Integer.MAX_VALUE
    }

    override fun visitPlaceholder(placeholderNode: PlaceholderNode): Int? {
        return 0
    }

    override fun visitBreak(breakNode: BreakNode): Int? {
        return 1
    }

    override fun visitContinue(continueNode: ContinueNode): Int? {
        return 1
    }

    override fun visitOverride(overrideNode: OverrideNode): Int? {
        return arguments(overrideNode).size
    }

    override fun visitClass(classNode: ClassNode): Int? {
        return arguments(classNode).size
    }

    override fun visitCatch(catchNode: CatchNode): Int? {
        return Integer.MAX_VALUE
    }
}
